core/shared/src/test/scala/cats/parse/ErrorShowTest.scala

```diff
@@ -0,0 +1,182 @@
+      loc: munit.Location
+  ): Unit = {
+    test(input) {
+      parser
+        .parseAll(input)
+        .fold(
+          e =>
+            assert(
+              e.show == expected,
+              s"""|obtained:
+              |--------
+              |${e.show}
+              |--------
+              |
+              |expected:
+              |$expected""".stripMargin
+            ),
+          _ => fail("should not parse")
+        )
+    }
+  }
+
+  val ok = string("ok")
+  val nl = string("\n")
+  val lx = string("l") ~ digits
+  val lxOk = ((lx | ok) ~ nl)
+
+  // # Expectations:
+  // OneOfStr
+  error(
+    string("foo") | string("bar") | string("baz"),
+    "ko",
+    s"""|ko
+        |^
+        |* one of: {"bar", "baz", "foo"}""".stripMargin
+  )
+
+  // InRange
+  error(
+    charIn(List('a', 'c', 'x', 'y')),
+    "ko",
+    """|ko
+       |^
+       |* is: a
+       |* is: c
+       |* in range: [x, y]""".stripMargin
+  )
+
+  // StartOfString
+  error(
+    ok ~ start,
+    "ok",
+    """|ok
+       |  ^
+       |* start of string""".stripMargin
+  )
+
+  // EndOfString
+  error(
+    lx.repExactlyAs[Int](2),
+    "l1l2l3".stripMargin,
+    """|l1l2l3
+       |    ^
+       |* end of string, length = 6""".stripMargin
+  )
+
+  // Length
+  error(
+    ok | length(2),
+    "aaaaaaaaaaaaaaaaaaa",
+    """|aaaaaaaaaaaaaaaaaaa
+       |  ^
+       |* end of string, length = 19""".stripMargin
+  ) // this does repport the correct error IMO
+
+  // ExpectedFailureAt
+  error(
+    not(ok),
+    "okidou",
+    """|okidou
+       |^
+       |* failure at ok""".stripMargin
+  )
+
+  // Fail
+  error(
+    ok ~ failWith("ko"),
+    "ok",
+    """|ok
+       |  ^
+       |* fail: ko""".stripMargin
+  )
+
+  // FailWith
+  error(
+    failWith("error msg"),
+    "ok",
+    """|ok
+       |^
+       |* fail: error msg""".stripMargin
+  )
+
+  // WithContext
+  error(
+    withContext0(ok, "ok") | withContext0(lx, "lx"),
+    "ko",
+    """|ko
+       |^
+       |* ok, one of: {"ok"}
+       |* lx, is: l""".stripMargin
+  )
+
+  // Context
+  error(
+    lxOk.rep(9),
+    """|l1
+       |l2
+       |l3
+       |l4
+       |ko
+       |l6
+       |l7
+       |l8
+       |l9
+       |""".stripMargin,
+    """|...
+       |l3
+       |l4
+       |ko
+       |^
+       |* one of: {"ok"}
+       |* is: l
+       |l6
+       |l7
+       |...""".stripMargin
+  )
+
+  error(
+    lxOk.rep(3),
+    """|l1
+       |ko
+       |l3""".stripMargin,
+    """|l1
+       |ko
+       |^
+       |* one of: {"ok"}
+       |* is: l
+       |l3""".stripMargin
+  )
+
+  error(
+    lxOk.rep(2),
+    """|l1
+       |ko""".stripMargin,
+    """|l1
+       |ko
+       |^
+       |* one of: {"ok"}
+       |* is: l""".stripMargin
+  )
+
+  error(
+    lxOk.rep(2),
+    """|ko
+       |l2""".stripMargin,
+    """|ko
+       |^
+       |* one of: {"ok"}
+       |* is: l
+       |l2""".stripMargin
+  )
```

Comment:
can we also add a scalacheck that does something simple like:
```scala
forAll { (p: Parser[Any], in: String) =>
  p.parseAll(in) match {
    case Right(_) => ()
    case Left(err) =>
      // make sure we exercise all the paths in show, and don't throw.
      assert(err.show ne null)
  }
}
```

Reply:
Can you give me more details, I'm not sure I understand.

Is the idea here to generate arbitrary parsers and arbitrary output to have full branch coverage?

How do I construct an `Arbitrary[Parser[Any]]` the search space must be huge.

Reply:
Yes, the idea is to ensure that all branches are exercised and we don't have some cases where even basic things fail.

https://github.com/typelevel/cats-parse/blob/e31eb1172ecb268f782c7f4c2845b905bcd1f849/core/shared/src/test/scala/cats/parse/ParserTest.scala#L602

there is how you generate an arbitrary parser. The space is large, but it is already implemented (and how I think we have had so few bugs).

Reply:
Hum, I get:

```
[error] /home/gui/foos/cats-parse/core/shared/src/test/scala/cats/parse/ErrorShowTest.scala:180:12: diverging implicit expansion for type org.scalacheck.Arbitrary[A]
[error] starting with method arbFunction0 in object Arbitrary
[error]     forAll { (p: Parser[Any], in: String) => 
[error]            ^
[error] one error found
```

Reply:
I think it is the `Any` there. Try `forAll { (p: Parser[Unit], in: String) =>`

Reply:
assuming you have imported the method I linked to.